### PR TITLE
Allow uploads for public base url users with editor role

### PR DESCRIPTION
**Pull request by djmaze**
## Change Summary

Allows uploads for public base url users with editors. This was not working anymore in the latest versions (and I am not sure it was ever implemented correctly).

This fix is probably not 100% correct, but I need a working docker image for my quickfix, but don't I am not able to create the image manully successfully, so I created this PR.

## Change type

- [ ] feat: (new feature for the user, not a new feature for build script)
- [x] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification

* Share a project via the public base url with editor role.
* Try to upload a file on an attachment field when using the base url.

### CODE REVIEW

**Comment by o1lab** [comment]
@djmaze : thanks for the PR. Please sign the CLA above.

**Review by wingkwong — CHANGES_REQUESTED** [review]
If I just use the API without passing auth token, I'd get `Bad Request` rather than `Unauthorized`. I'd suggest to check if `id` and `isPublicBase` in the first if-statement first.

File: packages/nocodb/src/lib/meta/api/attachmentApis.ts [hunk 1]
```diff
@@ -16,26 +16,33 @@ import Local from '../../v1-legacy/plugins/adapters/storage/Local';
 import { NC_ATTACHMENT_FIELD_SIZE } from '../../constants';
 
 const isUploadAllowed = async (req: Request, _res: Response, next: any) => {
-  if (!req['user']?.id) {
+  if (!req['user']) {
```
Comment (wingkwong) [inline packages/nocodb/src/lib/meta/api/attachmentApis.ts:19]: The test case is still failing. I believe you can simply do in this way.

```ts
  if (
    !req['user']?.id ||
    (req['user']?.isPublicBase &&
      !req['user'].roles?.includes(ProjectRoles.EDITOR))
  ) {
    NcError.unauthorized('Unauthorized');
  }

```

File: packages/nocodb/src/lib/meta/api/attachmentApis.ts [hunk 2]
```diff
@@ -16,26 +16,33 @@ import Local from '../../v1-legacy/plugins/adapters/storage/Local';
...
+            this.orWhere('roles', ProjectRoles.EDITOR);
+          })
+          .andWhere('fk_user_id', req['user'].id)
+          .first()))
+    )
+      return next();
+    // check public base url with editor role
+    else if (
+      req['user'].isPublicBase &&
+      req['user'].roles?.includes(ProjectRoles.EDITOR)
```
Comment (wingkwong) [inline packages/nocodb/src/lib/meta/api/attachmentApis.ts:45]: these can be removed

File: packages/nocodb/src/lib/meta/api/attachmentApis.ts (hunk 1 above)
Comment (djmaze) [reply packages/nocodb/src/lib/meta/api/attachmentApis.ts]: Not quite.. Then it fails if *either* the id is missing or the public url & role. But the success case is that *both* the id is missing as well as the public url & role is are there.

**Comment by djmaze** [comment]
It would really help if someone could tell me how to run *single tests* in the development environment. (I couldn't easily find out, but am not a mocha / JS expert). Running all tests takes *ages* on my machine, so I switched to just pushing stuff here first.

File: packages/nocodb/src/lib/meta/api/attachmentApis.ts (hunk 1 above)
Comment (wingkwong) [reply packages/nocodb/src/lib/meta/api/attachmentApis.ts]: Lemme push something to test.

**Comment by wingkwong** [comment]
> It would really help if someone could tell me how to run _single tests_ in the development environment. (I couldn't easily find out, but am not a mocha / JS expert). Running all tests takes _ages_ on my machine, so I switched to just pushing stuff here first.

You can simply add `.only` to the test case(s) that you want to run.

ref: https://docs.nocodb.com/engineering/unit-testing#walk-through-of-writing-a-unit-test

image

**Comment by djmaze** [comment]
@wingkwong Thanks!

**Pull request merged**